### ISSUE TITLE
[INFO] session and headers already sent

### ISSUE THREAD

**Issue by ShockedPlot7560** [upscalesoftware/swoole-session#4]
When starting a website conversion with swoole, I had to use this middleware to be able to use the sessions properly.  
Like many people using twig for rendering, I ran into the typical header already sent problem when working with sessions.

While trying to understand the origin with the #1  issue and searching on the internet, I could observe that nobody else had looked at the issue.

So, through this issue I would like to bring for future people, how I could solve my problem and why. Any information is good to take from other people.

> I can already say that I have had no problems with coroutine so far.

So to start with, I made a simple application like everyone else, with these few lines: 
```php
$server = new Swoole\HTTP\Server("localhost", 9501);

$server->on("WorkerStart", function($server, $workerId) {
    echo "Worker Start: $workerId\n";
});

$server->on("Start", function() {
	echo "Swoole HTTP Server Started @ 127.0.0.1:9501\n";
});

$server->on('Request', new SessionMiddleware(function (Swoole\Http\Request $request,Swoole\Http\Response $response) use ($app) {
    $response->end("Hello world !");
}));

$server->on("Shutdown", function($server, $workerId) {
	echo "Server shutting down...\n";
});

// Triggered when worker processes are being stopped
$server->on("WorkerStop", function($server, $workerId) {
	echo "Worker Stopped: $workerId\n";
});

$server->start();
```

A basic web server with swoole using SessionMiddleware. 

However, warnings will appear every time you want to access your site. Normal, we'll see about that.
I thought it was a problem on the ``on('Request')`` side, but not at all because we finish the request directly, so there is no need to print elsewhere.

After some research, I noticed that by removing the echo from ``WorkerStart``, everything worked fine. And this is quite logical, in fact, since they are the ones who will carry out the different requests, if we do an echo, the requests that we will process afterwards will be bugged as we do not want.

In conclusion, when rendering with any library, as a precaution use the output buffering to prevent any risk.
As for logging, rather than echoing or printing, use a lib to log like ``monolog`` under file to avoid any risk of writing to the buffer.

> And if we follow these precautions, I do not see why coroutine would not work, if nothing is written in the buffer.
> However, as soon as a PHP warning is thrown by the code and this one is displayed, the system will bug. It is therefore necessary to prevent any risk by removing warnings?

**Comment by sshymko** [comment]
@ShockedPlot7560 
Thanks for providing a useful info to anybody porting over "classical" apps to Swoole!

I'd recommend using STDERR for logging instead of STDOUT (used by `echo`), for example:
```php
$server->on('WorkerStart', function($server, $workerId) {
    fwrite(STDERR, "Worker Start: $workerId\n");
});
```

**Comment by ShockedPlot7560** [comment]
Using this type of channel is indeed a functional bypass. However, in a more rigorous approach, it may seem problematic (in case of flow redirection).  
But it can be another solution to the problem.

Hopefully, the information provided will help other people.
